sync-core/src/test/java/com/cloudant/sync/datastore/AttachmentStreamFactoryTest.java

```diff
@@ -0,0 +1,370 @@
+        File zippedPlainText = f("fixture/EncryptedAttachmentTest_plainText.gz");
+
+        File actualOutput = new File(datastore_manager_dir, "temp" + UUID.randomUUID());
+        OutputStream out = asf.getOutputStream(actualOutput, Attachment.Encoding.Gzip);
+
+        IOUtils.copy(new FileInputStream(plainText), out);
+        out.close();
+
+        Assert.assertTrue("Writing to unencrypted, encoded stream didn't give correct output",
+                IOUtils.contentEquals(
+                        new FileInputStream(actualOutput),
+                        new FileInputStream(zippedPlainText)));
+
+    }
+
+    @Test
+    /**
+     * Assert passing an AttachmentStreamFactory with no key reads an unencrypted and
+     * gzipped file correctly.
+     */
+    public void testSavedAttachmentReadsUnencryptedEncodedStream()
+            throws AttachmentException, IOException {
+
+        AttachmentStreamFactory asf = new AttachmentStreamFactory(new NullKeyProvider());
+
+        File plainText = f("fixture/EncryptedAttachmentTest_plainText");
+        File zippedPlainText = f("fixture/EncryptedAttachmentTest_plainText.gz");
+
+        SavedAttachment savedAttachment = new SavedAttachment("test", 0, 0, null, "text/plain",
+                zippedPlainText, Attachment.Encoding.Gzip, asf);
+
+        Assert.assertTrue("Reading unencrypted, encoded stream didn't give correct output",
+                IOUtils.contentEquals(
+                        savedAttachment.getInputStream(),
+                        new FileInputStream(plainText)));
+
+    }
+
+    // =====================
+    // Encrypted, un-encoded
+    // =====================
+
+    @Test
+    /**
+     * Assert passing an AttachmentStreamFactory with known key writes an encrypted file to disk.
+     */
+    public void testPreparedAttachmentWritesEncryptedUnencodedStream() throws AttachmentException,
+            IOException, InvalidKeyException {
+
+        AttachmentStreamFactory asf = new AttachmentStreamFactory(
+                EncryptionTestConstants.keyProvider16Byte
+        );
+
+        File plainText = f("fixture/EncryptedAttachmentTest_plainText");
+
+        UnsavedFileAttachment usf = new UnsavedFileAttachment(plainText, "text/plain");
+        PreparedAttachment preparedAttachment = new PreparedAttachment(
+                usf, datastore_manager_dir, asf);
+
+        // We check content using EncryptedAttachmentInputStream. It seems a bit circular,
+        // but the aim of this test is to make sure AttachmentStreamFactory is providing
+        // encrypted output readable by EncryptedAttachmentInputStream rather than exactly
+        // what that output is (EncryptedAttachmentOutputStream's own tests do that).
+        Assert.assertTrue("Writing to encrypted, un-encoded stream didn't give correct output",
+                IOUtils.contentEquals(
+                        new EncryptedAttachmentInputStream((
```

Comment:
Thought I'd got all those, but yup, think I finally have in 94e6878d0d38456663bfda630770b5924b510f9c